packages/flutter/lib/src/gestures/tap_and_drag.dart

```diff
@@ -1138,36 +1160,51 @@ sealed class BaseTapAndDragGestureRecognizer extends OneSequenceGestureRecognize
   @override
   String get debugDescription => 'tap_and_drag';
 
+  /// Calculates the update delta from the given pointer event.
+  OffsetPair _calculateDragUpdateDelta(PointerEvent event) {
+    final Offset localDelta = event.localDelta;
+
+    if (localDelta == Offset.zero) {
+      return OffsetPair.zero;
+    }
+
+    final Offset correctedLocalPosition = _initialPosition.local + localDelta;
+    final Matrix4? localToGlobalTransform = event.transform != null ? Matrix4.tryInvert(event.transform!) : null;
+    final Offset globalUpdateDelta = PointerEvent.transformDeltaViaPositions(
+      transform: localToGlobalTransform,
+      untransformedDelta: localDelta,
+      untransformedEndPosition: correctedLocalPosition,
+    );
+    return OffsetPair(local: localDelta, global: globalUpdateDelta);
+  }
+
+  void _updateCurrentPosition(OffsetPair delta) {
+    _currentPosition += delta;
+  }
+
   void _acceptDrag(PointerEvent event) {
     if (!_wonArenaForPrimaryPointer) {
       return;
     }
     if (dragStartBehavior == DragStartBehavior.start) {
-      _initialPosition = _initialPosition + OffsetPair(global: event.delta, local: event.localDelta);
+      _initialPosition += OffsetPair(global: event.delta, local: event.localDelta);
+      _currentPosition = _initialPosition;
     }
     _checkDragStart(event);
     if (event.localDelta != Offset.zero) {
-      final Matrix4? localToGlobal = event.transform != null ? Matrix4.tryInvert(event.transform!) : null;
-      final Offset correctedLocalPosition = _initialPosition.local + event.localDelta;
-      final Offset globalUpdateDelta = PointerEvent.transformDeltaViaPositions(
-        untransformedEndPosition: correctedLocalPosition,
-        untransformedDelta: event.localDelta,
-        transform: localToGlobal,
-      );
-      final OffsetPair updateDelta = OffsetPair(local: event.localDelta, global: globalUpdateDelta);
-      _correctedPosition = _initialPosition + updateDelta; // Only adds delta for down behaviour
-      _checkDragUpdate(event);
-      _correctedPosition = null;
+      final OffsetPair delta = _calculateDragUpdateDelta(event);
+      _updateCurrentPosition(delta);
+      _checkDragUpdate(event, corrected: _initialPosition + delta);
     }
   }
 
   void _checkDrag(PointerMoveEvent event) {
-    final Matrix4? localToGlobalTransform = event.transform == null ? null : Matrix4.tryInvert(event.transform!);
+    final Matrix4? localToGlobalTransform = event.transform != null ? Matrix4.tryInvert(event.transform!) : null;
```

Comment:
nit: prefer the shorter result on the left if possible.

Reply:
Sorry you mean `? null : Matrix4` is preferred?

Reply:
Yeah I think the original way this was written is preferred.

Reply:
Sure. Updated